File: source/java-sdk/src/main/java/com/aerospike/examples/advancedexpressions/AdvancedExpressions.java
package com.aerospike.examples.advancedexpressions;

import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicInteger;

import com.aerospike.client.sdk.Record;
import com.aerospike.client.sdk.Session;
import com.aerospike.client.sdk.TypedDataSet;
import com.aerospike.client.sdk.TypedKey;
import com.aerospike.client.sdk.TypedRecordStream;
import com.aerospike.client.sdk.cdt.ListOrder;
import com.aerospike.client.sdk.cdt.ListReturnType;
import com.aerospike.client.sdk.exp.Exp;
import com.aerospike.client.sdk.exp.ListExp;
import com.aerospike.examples.UseCase;
import com.aerospike.examples.advancedexpressions.model.Car;

/**
 * SDK port of the legacy {@code AdvancedExpressions} (see ../../java). Not a use case per se but
 * a set of techniques for advanced expression usage:
 * <ol>
 *   <li>Checking whether a value is in a list bin ({@code "Sunroof" IN features})</li>
 *   <li>Checking whether a bin's value is in a passed-in list ({@code color IN ["Red","Green","Blue"]})</li>
 *   <li>Performing multiple operations, some depending on others' results, within one operate call</li>
 * </ol>
 */
public class AdvancedExpressions implements UseCase {

    private static final int NUM_CARS = 10_000;

    private static final String[] MAKES = {"Toyota", "Holden", "Mitsubishi", "Kia", "Ferrari", "Volvo", "Audi", "Datsun", "Suzuki"};
    private static final String[] MODELS = {"Corolla", "Commodore", "Outlander", "Sorento", "Swift", "488 Spider", "XC90"};
    private static final String[] COLORS = {"Red", "Green", "Blue", "White", "Black", "Silver", "Purple"};
    private static final String[] ALL_FEATURES = {
            "Bluetooth Connectivity", "USB Charging Ports", "Apple CarPlay", "Android Auto", "Heated Seats",
            "Ventilated Seats", "Sunroof", "Panoramic Roof", "Navigation System", "Keyless Entry",
            "Push Button Start", "Remote Start", "Adaptive Cruise Control", "Blind Spot Monitoring",
            "Lane Departure Warning", "Automatic Emergency Braking", "Parking Sensors", "Rearview Camera",
            "360-Degree Camera", "Leather Upholstery", "Wireless Charging", "Heads-Up Display",
            "Premium Sound System", "LED Headlights", "Rain-Sensing Wipers", "Heated Steering Wheel",
            "Power Liftgate", "Roof Rails", "Tow Package", "Ambient Interior Lighting"
    };

    @Override
    public String getName() {
        return "Advanced Expressions";
    }

    @Override
    public String getDescription() {
        return "A set of techniques showing advanced usage of expressions. This is not a use case per-se but rather a set "
                + "of techniques that can be used in a use case. Current examples include:\n"
                + "1. Seeing if an item is in a list in the record in the database\n"
                + "2. Seeing if an item in the database is contained in a passed list\n"
                + "3. Performing multiple operations that return information in a single operation within one operate command.";
    }

    @Override
    public String getReference() {
        return "https://github.com/aerospike-examples/use-case-cookbook/blob/main/UseCases/advanced-expressions.md";
    }

    private final TypedDataSet<Car> cars =
            TypedDataSet.of(System.getProperty("demo.namespace", "test"), "uccb_car", Car.class);

    private Car randomCar(int id) {
        int numFeatures = ThreadLocalRandom.current().nextInt(0, 9);
        List<String> features = ThreadLocalRandom.current().ints(0, ALL_FEATURES.length)
                .distinct().limit(numFeatures).mapToObj(i -> ALL_FEATURES[i]).toList();
        Car.BodyType[] bodyTypes = Car.BodyType.values();
        return new Car(id,
                MAKES[ThreadLocalRandom.current().nextInt(MAKES.length)],
                MODELS[ThreadLocalRandom.current().nextInt(MODELS.length)],
                ThreadLocalRandom.current().nextInt(2000, 2025),
                bodyTypes[ThreadLocalRandom.current().nextInt(bodyTypes.length)].name(),
                ThreadLocalRandom.current().nextInt(7, 80) / 10.0,
                COLORS[ThreadLocalRandom.current().nextInt(COLORS.length)],
                ThreadLocalRandom.current().nextInt(0, 200_000),
                ThreadLocalRandom.current().nextInt(5_000, 150_000),
                features);
    }

    @Override
    public void setup(Session session) throws Exception {
        session.truncate(cars);

        System.out.printf("Generating %,d Cars%n", NUM_CARS);
        for (int id = 1; id <= NUM_CARS; id++) {
            session.upsert(cars).object(randomCar(id)).execute();
        }
    }

    @Override
    public void run(Session session) throws Exception {
        System.out.println("Find 10 cars which have Sunroofs. The features of the car are stored in the 'features' bin, so this is "
                + "effectively doing:\n"
                + "      \"Sunroof\" IN features\n"
                + "Note that this could be done using a secondary index too.");

        findCarsWithFeature(session, "Sunroof");

        System.out.println("\n\n");
        System.out.println("Find 10 cars whose color is any of Red, Green or Blue. Since Aerospike does not have an IN operation, "
                + "this will use a list operation to perform this. Effectively this is doing:\n"
                + "      color IN [\"Red\", \"Green\", \"Blue\"]");
        findCarsWithColors(session, List.of("Red", "Green", "Blue"));

        multipleCommandsInOneOperation(session);

        // The "acc" list-append is written as AEL (verified against a live cluster to behave
        // identically to ListExp.append) - per Tim's review comment to prefer AEL where possible.
        // The "counter" read-back is NOT converted: every AEL index-read syntax tried against a live
        // cluster ($.acc.getByIndex(0), $.acc[0], $.acc.{0}, $.acc.{0}.getValue()) threw the same
        // server-side "Parameter error" that blocked Leaderboard's map-index read (see that class's
        // getScoresAroundPlayer javadoc) - single-element list/map index reads via AEL appear to be
        // an unsupported composition on this build, not just a syntax guess away. Kept on the
        // proven ListExp.getByIndex Exp form; worth revisiting with correct syntax from Tim.
        TypedKey<Car> key = cars.id(1);
        session.upsert(key)
                .bin("acc").listCreate(ListOrder.UNORDERED)
                .bin("acc").upsertFrom("$.acc.append(10)")
                .bin("counter").upsertFrom(ListExp.getByIndex(ListReturnType.VALUE, Exp.Type.INT, Exp.val(0), Exp.listBin("acc")))
                .bin("acc").remove()
                .execute();
    }

    /**
     * Kept as {@code ListExp.getByValue(EXISTS, ...)} rather than an AEL string per Tim's review
     * comment: tried two AEL translations of "value exists in a list bin" against a live cluster
     * ({@code $.features.contains('Sunroof')} - server-side {@code Parameter error}; {@code
     * $.features.{='Sunroof'}.count() > 0} - parses fine but silently returns 0 matches instead of
     * the correct 10) and neither is safe to ship without an authoritative AEL grammar reference
     * for list-membership/selector syntax. Worth revisiting with the correct syntax from Tim.
     */
    private void findCarsWithFeature(Session session, String feature) {
        Exp exp = ListExp.getByValue(ListReturnType.EXISTS, Exp.val(feature), Exp.listBin("features"));
        showCarsMatchingExpression(session, exp, 10);
    }

    /** Same rationale as {@link #findCarsWithFeature} - kept as Exp, not converted to AEL. */
    private void findCarsWithColors(Session session, List<String> colors) {
        Exp exp = ListExp.getByValue(ListReturnType.EXISTS, Exp.stringBin("color"), Exp.val(colors));
        showCarsMatchingExpression(session, exp, 10);
    }

    /**
     * Using just ONE operation, adds:
     * <ul>
     *   <li>a feature "Great Color" if the color is Purple</li>
     *   <li>"Looks Cool" if the body type is CONVERTIBLE</li>
     *   <li>"Powerful" if the engine is over 5L</li>
     *   <li>"New-ish" if the year is 2020 or later</li>
     * </ul>
     */
    private void multipleCommandsInOneOperation(Session session) {
        TypedKey<Car> key = cars.id(1);
        session.upsert(key).bin("color").setTo("Purple").execute();
        System.out.println("Record before augmenting:");
        showCar(1, session.query(key).execute().getFirstRecord());

        // Written as AEL rather than nested Exp.let/Exp.def/Exp.cond builder calls - much easier to
        // read once the nesting gets this deep, and it's the same expression the server evaluates.
        session.upsert(key)
                .bin("features").upsertFrom("""
                        let (
                          color = when (
                            $.color == 'Purple' => $.features.append('Great Color'),
                            default => $.features
                          ),
                          type = when (
                            $.bodyType == 'CONVERTIBLE' => (${color}).append('Looks Cool'),
                            default => ${color}
                          ),
                          power = when (
                            $.engineSize > 5.0 => (${type}).append('Powerful'),
                            default => ${type}
                          ),
                          age = when (
                            $.year >= 2020 => (${power}).append('New-ish'),
                            default => ${power}
                          )
                        ) then (${age})
                        """)
                .execute();
        System.out.println("Record after augmenting:");
        showCar(1, session.query(key).execute().getFirstRecord());
    }

    private void showCarsMatchingExpression(Session session, Exp exp, int limit) {
        AtomicInteger count = new AtomicInteger();
        try (TypedRecordStream<Car> stream = session.query(cars).where(exp).limit(limit).execute()) {
            stream.forEach(result -> {
                if (result.isOk()) {
                    showCar(count.incrementAndGet(), result.recordOrThrow());
                }
            });
        }
    }

    private void showCar(int count, Record rec) {
        System.out.printf("%nCar %d (id: %d)%n", count, rec.getInt("id"));
        System.out.printf("\tMake:        %s%n", rec.getString("make"));
        System.out.printf("\tModel:       %s%n", rec.getString("model"));
        System.out.printf("\tYear:        %s%n", rec.getInt("year"));
        System.out.printf("\tColor:       %s%n", rec.getString("color"));
        System.out.printf("\tType:        %s%n", rec.getString("bodyType"));
        System.out.printf("\tEngine Size: %s%n", rec.getDouble("engineSize"));
        System.out.printf("\tFeatures:    %s%n", rec.getList("features"));
    }
}

File: source/python-sdk/usecasecookbook/advancedexpressions/advanced_expressions.py
"""Port of ../../java-sdk's AdvancedExpressions (itself an SDK port of ../../java). Not a
use case per se but a set of techniques for advanced expression usage:

1. Checking whether a value is in a list bin (``"Sunroof" IN features``)
2. Checking whether a bin's value is in a passed-in list (``color IN ["Red","Green","Blue"]``)
3. Performing multiple operations, some depending on others' results, within one write

All three port to AEL exactly as ../../java-sdk has them - including technique 3's single
nested ``let``/``when`` write with ``append(value)`` branches, which requires server-side AEL
compilation (Aerospike 8.2.0+; see ``../README.md``).
"""

import random

from aerospike_async import ListOrderType
from aerospike_sdk import DataSet, QueryHint
from aerospike_sdk.sync import Session

from usecasecookbook import config
from usecasecookbook.advancedexpressions.model import Car
from usecasecookbook.use_case import UseCase

NUM_CARS = 10_000

MAKES = ["Toyota", "Holden", "Mitsubishi", "Kia", "Ferrari", "Volvo", "Audi", "Datsun", "Suzuki"]
MODELS = ["Corolla", "Commodore", "Outlander", "Sorento", "Swift", "488 Spider", "XC90"]
COLORS = ["Red", "Green", "Blue", "White", "Black", "Silver", "Purple"]
BODY_TYPES = ["SEDAN", "SUV", "HATCHBACK", "COUPE", "CONVERTIBLE", "UTE"]
ALL_FEATURES = [
    "Bluetooth Connectivity", "USB Charging Ports", "Apple CarPlay", "Android Auto", "Heated Seats",
    "Ventilated Seats", "Sunroof", "Panoramic Roof", "Navigation System", "Keyless Entry",
    "Push Button Start", "Remote Start", "Adaptive Cruise Control", "Blind Spot Monitoring",
    "Lane Departure Warning", "Automatic Emergency Braking", "Parking Sensors", "Rearview Camera",
    "360-Degree Camera", "Leather Upholstery", "Wireless Charging", "Heads-Up Display",
    "Premium Sound System", "LED Headlights", "Rain-Sensing Wipers", "Heated Steering Wheel",
    "Power Liftgate", "Roof Rails", "Tow Package", "Ambient Interior Lighting",
]

CARS = DataSet.of(config.NAMESPACE, "uccb_car")


def _random_car(car_id: int) -> Car:
    num_features = random.randint(0, 8)
    features = random.sample(ALL_FEATURES, num_features)
    return Car(
        id=car_id,
        make=random.choice(MAKES),
        model=random.choice(MODELS),
        year=random.randint(2000, 2024),
        body_type=random.choice(BODY_TYPES),
        engine_size=random.randint(7, 79) / 10.0,
        color=random.choice(COLORS),
        milage=random.randint(0, 199_999),
        price=random.randint(5_000, 149_999),
        features=features,
    )


class AdvancedExpressions(UseCase):
    def get_name(self) -> str:
        return "Advanced Expressions"

    def get_description(self) -> str:
        return (
            "A set of techniques showing advanced usage of expressions. This is not a use case per-se but rather a "
            "set of techniques that can be used in a use case. Current examples include:\n"
            "1. Seeing if an item is in a list in the record in the database\n"
            "2. Seeing if an item in the database is contained in a passed list\n"
            "3. Performing multiple operations that return information in a single operation within one operate command."
        )

    def get_reference(self) -> str:
        return "https://github.com/aerospike-examples/use-case-cookbook/blob/main/UseCases/advanced-expressions.md"

    def setup(self, session: Session) -> None:
        session.truncate(CARS)

        print(f"Generating {NUM_CARS:,} Cars")
        for car_id in range(1, NUM_CARS + 1):
            car = _random_car(car_id)
            session.upsert(CARS.id(car.id)).put(car.to_bins()).execute()

    def run(self, session: Session) -> None:
        print(
            "Find 10 cars which have Sunroofs. The features of the car are stored in the 'features' bin, so this is "
            "effectively doing:\n"
            "      \"Sunroof\" IN features\n"
            "Note that this could be done using a secondary index too."
        )
        self._find_cars_with_feature(session, "Sunroof")

        print("\n\n")
        print(
            "Find 10 cars whose color is any of Red, Green or Blue. Since Aerospike does not have an IN operation, "
            "this will use a list operation to perform this. Effectively this is doing:\n"
            "      color IN [\"Red\", \"Green\", \"Blue\"]"
        )
        self._find_cars_with_colors(session, ["Red", "Green", "Blue"])

        self._multiple_commands_in_one_operation(session)

        # AEL's append(value) can't create a missing list bin, so it's created via the native
        # CDT builder first; the append and the typed read-back both use AEL.
        key = CARS.id(1)
        session.upsert(key).bin("acc").list_create(ListOrderType.UNORDERED).execute()
        session.upsert(key).bin("acc").upsert_from("$.acc.append(10)").execute()
        session.upsert(key).bin("counter").upsert_from("$.acc.[0]:INT").execute()
        session.upsert(key).bin("acc").remove().execute()

    def _find_cars_with_feature(self, session: Session, feature: str) -> None:
        """Membership check via AEL's ``value in $.bin`` operator."""
        self._show_cars_matching_expression(session, f"'{feature}' in $.features", 10)

    def _find_cars_with_colors(self, session: Session, colors: list[str]) -> None:
        """Same operator, reversed direction: ``$.color in [...]``."""
        color_list = ", ".join(f'"{c}"' for c in colors)
        self._show_cars_matching_expression(session, f"$.color in [{color_list}]", 10)

    def _multiple_commands_in_one_operation(self, session: Session) -> None:
        """Adds all 4 conditional features in ONE write via a 4-level nested AEL ``let``/``when``
        expression, matching ../../java-sdk's ``multipleCommandsInOneOperation``.
        """
        key = CARS.id(1)
        session.upsert(key).bin("color").set_to("Purple").execute()

        print("Record before augmenting:")
        self._print_car(1, self._read_car(session, key).to_bins())

        ael = """
            let (
              color = when ($.color == 'Purple' => $.features.append('Great Color'), default => $.features),
              type  = when ($.bodyType == 'CONVERTIBLE' => (${color}).append('Looks Cool'), default => ${color}),
              power = when ($.engineSize > 5.0 => (${type}).append('Powerful'), default => ${type}),
              age   = when ($.year >= 2020 => (${power}).append('New-ish'), default => ${power})
            ) then (${age})
        """
        session.upsert(key).bin("features").upsert_from(ael).execute()

        print("Record after augmenting:")
        self._print_car(1, self._read_car(session, key).to_bins())

    @staticmethod
    def _read_car(session: Session, key) -> Car:
        stream = session.query(key).execute()
        car = None
        for row in stream:
            if row.is_ok and row.record is not None:
                car = Car.from_bins(row.record.bins)
        stream.close()
        assert car is not None, f"no car found for {key}"
        return car

    def _show_cars_matching_expression(self, session: Session, ael: str, limit: int) -> None:
        # No secondary index on "features"/"color" - this is an intentional full-set scan (the
        # use case's own description notes a secondary index could do this instead), so opt into
        # the primary-index fallback explicitly rather than have the server reject it.
        stream = (
            session.query(CARS)
            .where(ael)
            .with_hint(QueryHint(allow_scans_with_where=True))
            .limit(limit)
            .execute()
        )
        count = 0
        for row in stream:
            if row.is_ok and row.record is not None:
                count += 1
                self._print_car(count, row.record.bins)
        stream.close()

    @staticmethod
    def _print_car(count: int, bins: dict) -> None:
        print()
        print(f"Car {count} (id: {bins['id']})")
        print(f"\tMake:        {bins['make']}")
        print(f"\tModel:       {bins['model']}")
        print(f"\tYear:        {bins['year']}")
        print(f"\tColor:       {bins['color']}")
        print(f"\tType:        {bins['bodyType']}")
        print(f"\tEngine Size: {bins['engineSize']}")
        print(f"\tFeatures:    {bins['features']}")
